build_common.ps1

```diff
@@ -84,47 +91,88 @@ class BuildProject {
 			$this._LoadContentOptions()
 			$this._ValidateProjectFiles()
-			$this._CleanAdditional()
-			$this._CopyModToSdk()
-			$this._ConvertLocalization()
-			$this._CopyToSrc()
-			$this._RunPreMakeHooks()
-			$this._CheckCleanCompiled()
-			$this._RunMakeBase()
-			$this._RunMakeMod()
+			$this._PerformStep({ ($_)._CleanAdditional() }, "Cleaning", "Cleaned", "additional mods")
+			$this._PerformStep({ ($_)._CopyModToSdk() }, "Mirroring", "Mirrored", "mod to SDK")
+			$this._PerformStep({ ($_)._ConvertLocalization() }, "Converting", "Converted", "Localization UTF-8 -> UTF-16")
+			$this._PerformStep({ ($_)._CopyToSrc() }, "Populating", "Populated", "Development\Src folder")
+			$this._PerformStep({ ($_)._RunPreMakeHooks() }, "Running", "Ran", "Pre-Make hooks")
+			$this._PerformStep({ ($_)._CheckCleanCompiled() }, "Verifying", "Verified", "compiled script packages")
+			$this._PerformStep({ ($_)._RunMakeBase() }, "Compiling", "Compiled", "base-game script packages")
+			$this._PerformStep({ ($_)._RunMakeMod() }, "Compiling", "Compiled", "mod script packages")
 			$this._RecordCoreTimestamp()
 			if ($this.isHl) {
 				if (-not $this.debug) {
-					$this._RunCookHL()
+					$this._PerformStep({ ($_)._RunCookHL() }, "Cooking", "Cooked", "Highlander packages")
 				} else {
 					Write-Host "Skipping cooking as debug build"
 				}
 			}
-			$this._CopyScriptPackages()
+			$this._PerformStep({ ($_)._CopyScriptPackages() }, "Copying", "Copied", "compiled script packages")
 			
 			# The shader step needs to happen before cooking - precompiler gets confused by some inlined materials
-			$this._PrecompileShaders()
+			$this._PerformStep({ ($_)._PrecompileShaders() }, "Precompiling", "Precompiled", "shaders")
 	
-			$this._RunCookAssets()
+			$this._PerformStep({ ($_)._RunCookAssets() }, "Cooking", "Cooked", "mod assets")
 	
 			# Do this last as there is no need for it earlier - the cooker obviously has access to the game assets
 			# and precompiling shaders seems to do nothing (I assume they are included in the game's GlobalShaderCache)
-			$this._CopyMissingUncooked()
+			$this._PerformStep({ ($_)._CopyMissingUncooked() }, "Copying", "Copied", "requested uncooked packages")
 	
-			$this._FinalCopy()
-
-			SuccessMessage "*** SUCCESS! ***" $this.modNameCanonical
+			$this._PerformStep({ ($_)._FinalCopy() }, "Copying", "Copied", "built mod to game directory")
+			$fullStopwatch.Stop()
+			$this._ReportTimings($fullStopwatch)
+			SuccessMessage "*** SUCCESS! ($(FormatElapsed $fullStopwatch.Elapsed)) ***" $this.modNameCanonical
 		}
 		catch {
 			[System.Media.SystemSounds]::Hand.Play()
 			throw
 		}
 	}
 
+	[void]_PerformStep([scriptblock]$stepCallback, [string]$progressWord, [string]$completedWord, [string]$description) {
+		Write-Host "$($progressWord) $($description)..."
+		$sw = [Diagnostics.Stopwatch]::StartNew()
+
+		# HACK: Set $_ for $stepCallback with Foreach-Object on only one object
+		$this | ForEach-Object $stepCallback
```

Comment:
Can't you do something like this instead?

```ps1
$that = $this
$this._PerformStep({ $that._DoStuff() })
```

Reply:
Action blocks still have a separate scope, so the only option is `{ ($using:this)._DoStuff() } `, which I found unnecessarily verbose. Let me know if you prefer that.

Reply:
Ah my bad, I thought the issue was with `$this` being set to different things (like in js). Let's keep the current approach